internal/pkg/kubernetes/client.go

```diff
@@ -0,0 +1,360 @@
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var (
+	typeMeta     meta.TypeMeta
+	ctx          context.Context
+	cancel       context.CancelFunc
+	timeoutError = errors.New("timeout when waiting job to be available")
+)
+
+func init() {
+	typeMeta = meta.TypeMeta{
+		Kind:       "Job",
+		APIVersion: "batch/v1",
+	}
+}
+
+type KubeClient interface {
+	ExecuteJob(jobId string, imageName string, envMap map[string]string) (string, error)
+	ExecuteJobWithCommand(jobId string, imageName string, args map[string]string, commands []string) (string, error)
+	JobExecutionStatus(executionName string) (string, error)
+	GetPodLogs(pod *v1.Pod) (io.ReadCloser, error)
+	WaitForReadyJob(executionName string, waitTime time.Duration) error
+	WaitForReadyPod(executionName string, waitTime time.Duration) (*v1.Pod, error)
+}
+
+type kubeClient struct {
+	clientSet                    kubernetes.Interface
+	namespace                    string
+	kubeServiceAccountName       string
+	jobPodAnnotations            map[string]string
+	kubeJobActiveDeadlineSeconds int
+	kubeJobRetries               int
+	kubeWaitForResourcePollCount int
+}
+
+func NewClientSet(kubernetesConfig config.OctaviusExecutorConfig) (*kubernetes.Clientset, error) {
+	var kubeConfig *rest.Config
+	if kubernetesConfig.KubeConfig == "out-of-cluster" {
+
+		home := os.Getenv("HOME")
+		kubeConfigPath := filepath.Join(home, ".kube", "kubeconfig")
+
+		configOverrides := &clientcmd.ConfigOverrides{}
+		if kubernetesConfig.KubeContext != "default" {
+			configOverrides.CurrentContext = kubernetesConfig.KubeContext
+		}
+
+		var err error
+		kubeConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeConfigPath},
+			configOverrides).ClientConfig()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		var err error
+		kubeConfig, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	clientSet, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientSet, nil
+}
+
+func NewKubernetesClient(kubernetesConfig config.OctaviusExecutorConfig) (KubeClient, error) {
+	newClient := &kubeClient{
+		namespace:                    kubernetesConfig.DefaultNamespace,
+		kubeServiceAccountName:       kubernetesConfig.KubeServiceAccountName,
+		jobPodAnnotations:            kubernetesConfig.JobPodAnnotations,
+		kubeJobActiveDeadlineSeconds: kubernetesConfig.KubeJobActiveDeadlineSeconds,
+		kubeJobRetries:               kubernetesConfig.KubeJobRetries,
+		kubeWaitForResourcePollCount: kubernetesConfig.KubeWaitForResourcePollCount,
+	}
+
+	var err error
+	newClient.clientSet, err = NewClientSet(kubernetesConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return newClient, nil
+}
+
+func jobLabel(executionName string) map[string]string {
+	return map[string]string{
+		"job": executionName,
```

Comment:
Try this convention while setting up kubernetes label
https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/